store/index.js

```diff
@@ -12,77 +51,140 @@ export const useMainStore = defineStore({
     },
+
     loadMagicItems() {
-      axios
-        .get(
-          `${
-            useRuntimeConfig().public.apiUrl
-          }/magicitems/?fields=slug,name,type,rarity,document__slug,document__title&limit=1000`
-        )
-        .then((response) => {
-          this.magicItemsList = response.data.results;
-        });
+      this.loadFromApi(
+        'magicitems',
+        magicItemFields,
+        5000,
+        magicItemOrder,
+        'magicItemsList'
+      );
     },
+
     loadBackgrounds() {
-      axios
-        .get(`${useRuntimeConfig().public.apiUrl}/backgrounds/?limit=1000`)
-        .then((response) => {
-          this.backgrounds = response.data.results;
-        });
+      this.loadFromApi('backgrounds', '', 1000, '', 'backgrounds');
     },
+
     loadClasses() {
-      axios
-        .get(`${useRuntimeConfig().public.apiUrl}/classes/`) //you will need to enable CORS to make this work
-        .then((response) => {
-          this.classes = response.data.results;
-        });
+      this.loadFromApi('classes', '', 1000, '', 'classes');
     },
+
     loadRaces() {
-      axios
-        .get(`${useRuntimeConfig().public.apiUrl}/races/`) //you will need to enable CORS to make this work
-        .then((response) => {
-          this.races = response.data.results;
-        });
+      this.loadFromApi('races', '', 1000, '', 'races');
     },
+
     loadSections() {
-      axios
-        .get(`${useRuntimeConfig().public.apiUrl}/sections/`) //you will need to enable CORS to make this work
-        .then((response) => {
-          this.sections = response.data.results;
-        });
+      this.loadFromApi('sections', '', 1000, '', 'sections');
+    },
+
+    loadDocuments() {
+      this.loadFromApi('documents', '', 1000, '', 'documents');
+    },
+    loadSourcesFromLocal() {
+      if (process.client) {
+        const savedSources = localStorage.getItem('sources');
+        return savedSources ? JSON.parse(savedSources) : [];
+      }
+    },
+    saveSourcesToLocal(sources) {
+      localStorage.setItem('sources', JSON.stringify(sources));
+    },
+    setSources(sources) {
+      if (this.sourceSelection === sources) {
+        return; // if the sources are the same, don't do anything
+      }
+      this.sourceSelection = sources;
+      this.saveSourcesToLocal(sources); // save to localStorage
+      this.sourceString = sources
+        ? `&document__slug__in=${this.sourceSelection.join(',')}` // if sources are selected, construct a query string segment for them
+        : '';
+      this.clearFresh(); // clear the list of fresh sources, since they are now stale
+    },
+    // load sources from the
+    async initializeSources() {
+      const savedSources = this.loadSourcesFromLocal();
+      if (savedSources.length > 0) {
+        this.setSources(savedSources);
+      } else {
+        await this.loadDocuments();
+        if (this.documents && this.documents.length > 0) {
```

Comment:
I spun up this branch and the development server was throwing a TypeError. Stack trace suggests that it could well be this conditional that is triggering it.

<img width="976" src="https://github.com/open5e/open5e/assets/47755775/054ad2a7-722c-48d1-b76e-f6d7ae9ce43d">

(NB. this warning does not appear on the `staging` branch)

This error doesn't appear to be breaking anything (yet), the source selection still works, but we should fix it.

It looks like the problem is that `this.documents` is undefined once we reach this point and so it doesn't have a `length` property. But as the source selector still works, so I guess that this is probably due to it trying to access the property before the promise is resolved, and trying again once it resolves? Perhaps we need to declare`loadDocuments()` as an async function?

If this doesn't work, perhaps we could have `loadDocuments()` additionally return the docs as an array and then use that inside the conditional

```const docs = this.loadDocuments();```
```if (docs.length > 0) {  ```
```  ...```